### PR TITLE
Add support for compute fleet status transitions

### PR DESCRIPTION
**Pull request by demartinofra**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/slurm_plugin/clustermgtd.py [hunk 1]
```diff
@@ -596,36 +713,38 @@ def _terminate_orphaned_instances(self, cluster_instances, ips_used_by_slurm):
...
-                instances_to_terminate, terminate_batch_size=self.sync_config.terminate_max_batch_size
+            self._instance_manager.delete_instances(
+                instances_to_terminate, terminate_batch_size=self._config.terminate_max_batch_size
             )
 
 
 def _run_clustermgtd():
     """Run clustermgtd actions."""
-    cluster_manager = ClusterManager()
+    config = ClustermgtdConfig(os.path.join(CONFIG_FILE_DIR, "parallelcluster_clustermgtd.conf"))
```
Comment (rexcsn) [inline src/slurm_plugin/clustermgtd.py:728]: Let's remove the `__repl__` call in `ClustemgtdConfig.__init__`, and have manage_cluster log the config so we don't have 2 logs of config at the start?

File: src/common/schedulers/slurm_commands.py [hunk 2]
```diff
@@ -151,6 +162,28 @@ def update_nodes(
...
+
+
+def update_all_partitions(state):
+    try:
+        partitions = [partition.name for partition in get_partition_info()]
+        succeeded_partitions = update_partitions(partitions, state)
+        return succeeded_partitions == partitions
+    except Exception as e:
+        logging.error("Failed when updating partitions with error %s", e)
+        return False
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply src/common/schedulers/slurm_commands.py:184]: Handling the logic of _manage_compute_fleet_status_transitions with an exception based approach makes the entire flow much more complex and hard to follow. I could raise an exception from here but then I'd still build a wrapper function to catch it and return a boolean, which is why I just kept a single function in the end. Let me know what you think.

File: src/slurm_plugin/common.py [hunk 3]
```diff
@@ -310,17 +310,26 @@ def get_cluster_instances(self, include_master=False, alive_states_only=True):
             for instance_info in filtered_iterator
         ]
 
+    def terminate_all_compute_nodes(self, terminate_batch_size):
+        try:
+            compute_nodes = self.get_cluster_instances()
+            self.delete_instances([instance.id for instance in compute_nodes], terminate_batch_size)
+            return True
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply src/slurm_plugin/common.py:317]: Same reason as the other function, happy to find a better alternative if needed.

File: src/slurm_plugin/clustermgtd.py [hunk 4]
```diff
@@ -178,43 +247,90 @@ class EC2InstancesInfoUnavailable(Exception):
...
+                if partitions_activated_successfully:
+                    self._update_compute_fleet_status(ComputeFleetStatus.RUNNING)
+        except ComputeFleetStatusManager.ConditionalStatusUpdateFailed:
+            log.warning(
+                "Cluster status was updated while handling a transition from %s. "
+                "Retrying the status transition in %d seconds",
+                self._compute_fleet_status,
+                LOOP_TIME,
+            )
+            time.sleep(LOOP_TIME)
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply src/slurm_plugin/clustermgtd.py]: the main reason was to not complicate even more the code in manage_cluster function, but I think you are right better not to sleep outside of the main loop

File: src/slurm_plugin/clustermgtd.py [hunk 5]
```diff
@@ -41,6 +46,68 @@
 log = logging.getLogger(__name__)
 
 
+class ComputeFleetStatus(Enum):
+    STOPPED = "STOPPED"
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply src/slurm_plugin/clustermgtd.py]: Done

File: src/slurm_plugin/clustermgtd.py [hunk 6]
```diff
@@ -178,43 +247,90 @@ class EC2InstancesInfoUnavailable(Exception):
...
+        log.info("Updating compute fleet status from %s to %s", self._compute_fleet_status, status)
+        self._compute_fleet_status_manager.update_status(current_status=self._compute_fleet_status, next_status=status)
+        self._compute_fleet_status = status
+
+    @log_exception(log, "handling compute fleet status transitions", catch_exception=Exception, raise_on_error=False)
+    def _manage_compute_fleet_status_transitions(self):
+        self._compute_fleet_status = self._compute_fleet_status_manager.get_status(fallback=self._compute_fleet_status)
+        log.info("Current compute fleet status: %s", self._compute_fleet_status)
+        try:
+            if ComputeFleetStatus.is_stop_status(self._compute_fleet_status):
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply src/slurm_plugin/clustermgtd.py:313]: Tried to better document the behavior. Let me know if it's enough

File: src/slurm_plugin/clustermgtd.py (hunk 1 above)
Comment (demartinofra) [reply src/slurm_plugin/clustermgtd.py:747]: done! I also slightly optimized this logic so that it checks if the new config is different from the previous one before re-initializing the objects and writing a log message.

File: src/slurm_plugin/clustermgtd.py [hunk 7]
```diff
@@ -74,10 +142,20 @@ def __repr__(self):
         attrs = ", ".join(["{key}={value}".format(key=key, value=repr(value)) for key, value in self.__dict__.items()])
         return "{class_name}({attrs})".format(class_name=self.__class__.__name__, attrs=attrs)
 
+    def __eq__(self, other):
+        if type(other) is type(self):
+            return self._config == other._config
```
Comment (rexcsn) [inline src/slurm_plugin/clustermgtd.py:147]: nit: Does comparing 2 ConfigParser object "just works"? I could not find the documentation, please double check

**Comment by rexcsn** [comment]
Note: There is a warning: `The base branch requires all commits to be signed`. Verified that commit in this PR is signed. Minor comment can be addressed later, merging PR

File: src/slurm_plugin/clustermgtd.py (hunk 7 above)
Comment (demartinofra) [reply src/slurm_plugin/clustermgtd.py:147]: I did test it manually but let me add a unit test for it: https://github.com/aws/aws-parallelcluster-node/pull/229
